src/eShop.ServiceDefaults/OpenApi.Extensions.cs

```diff
@@ -21,74 +21,114 @@ public static IApplicationBuilder UseDefaultOpenApi(this WebApplication app)
+    </head>
+    <body>
+        <div id="swagger-ui"></div>
+
+        <script src="https://unpkg.com/swagger-ui-dist/swagger-ui-standalone-preset.js"></script>
+        <script src="https://unpkg.com/swagger-ui-dist/swagger-ui-bundle.js"></script>
+
+        <script>
+            window.onload = function() {
+                const ui = SwaggerUIBundle({
+                    url: "/openapi/{{documentName}}.json",
+                        dom_id: '#swagger-ui',
+                        deepLinking: true,
+                        presets: [
+                            SwaggerUIBundle.presets.apis,
+                            SwaggerUIStandalonePreset
+                        ],
+                        plugins: [
+                            SwaggerUIBundle.plugins.DownloadUrl
+                        ],
+                        layout: "StandaloneLayout",
+                });
+                {{InitializeOAuth(authConfigurationSection)}}
+                window.ui = ui
+            }
+        </script>
+    </body>
+    </html>
+    """, "text/html")).ExcludeFromDescription();
+
+        static string InitializeOAuth(IConfigurationSection configurationSection)
+        {
+            if (configurationSection.Exists())
+            {
+                var clientId = configurationSection.GetRequiredValue("ClientId");
+                var appName = configurationSection.GetRequiredValue("AppName");
+                return $@"
+                    ui.initOAuth({{
+                            appName: ""{appName}"",
+                            clientId: ""{clientId}"",
+                            appName: ""{appName}"",
+                        }});
+                    ";
+            }
+            return string.Empty;
+        }
+    }
+
     public static IHostApplicationBuilder AddDefaultOpenApi(
         this IHostApplicationBuilder builder,
         IApiVersioningBuilder? apiVersioning = default)
     {
-        var services = builder.Services;
-        var configuration = builder.Configuration;
-        var openApi = configuration.GetSection("OpenApi");
+        var openApi = builder.Configuration.GetSection("OpenApi");
 
         if (!openApi.Exists())
         {
             return builder;
         }
 
-        services.AddEndpointsApiExplorer();
-
         if (apiVersioning is not null)
         {
             // the default format will just be ApiVersion.ToString(); for example, 1.0.
             // this will format the version as "'v'major[.minor][-status]"
-            apiVersioning.AddApiExplorer(options => options.GroupNameFormat = "'v'VVV");
-            services.AddTransient<IConfigureOptions<SwaggerGenOptions>, ConfigureSwaggerOptions>();
-            services.AddSwaggerGen(options => options.OperationFilter<OpenApiDefaultValues>());
+            var versioned = apiVersioning.AddApiExplorer(options => options.GroupNameFormat = "'v'VVV");
```

Comment:
@commonsensesoftware I think this does the trick WRT to applying the eager resolution pattern we talked about.

Reply:
@captainsafia mechanically, this looks good to me. You'll know that everything is working correctly if `IApiVersionDescriptionProvider` is picking up the collated set of API versions. To be sure, I would temporarily add (or change) an `Endpoint` to be something other than the default; for example, `2.0`. Double-check that it shows up. If it does, then you _should_ be good. When there are no collated API versions at all (which _shouldn't_ happen), the `IApiVersionDescriptionProvider` will always still include the value from `ApiVersioningOptions.DefaultApiVersion`, which is `ApiVersion.Default` (e.g. `1.0`) if otherwise unchanged. This can sometimes lead to a false sense of correctness if there is only one current API version defined and it happens to be `1.0`.

Reply:
Yep, I was a little too trigger-happy to declare that this works. A more thorough validation that including an additional API version proves that this doesn't work.

When the enumerate the `IApiVersionDescriptionProvider` upfront at startup, the mapped API versions are not yet registered so the eager resolution on the builder doesn't catch these values. This strategy works for Swashbuckle because the document name doesn't play a key role in the resolution of OpenAPI-related services like it does in the built-in implementation.

There's a bit of a chicken-egg problem to unravel here. `AddOpenApi` needs to know all the version names ahead of time in order to configure the keyed service dependencies correctly. `Asp.Versioning` doesn't resolve the full set of supported versions until _after_ the endpoints are mapped.

An "ugly" workaround is to require that users explicitly map the supported version numbers themselves in their startup logic:

```csharp
using Asp.Versioning.Builder;

var builder = WebApplication.CreateBuilder(args);

builder.Services
	.AddOpenApi("v1")
	.AddOpenApi("v2");

var app = builder.Build();

app.MapOpenApi();

app.NewVersionedApi("Catalog")
   .MapCatalogApiV1()
   .MapCatalogApiV2();

app.Run();
```